pandas/core/groupby/grouper.py

```diff
@@ -527,13 +520,10 @@ def __init__(
                 self.grouping_vector = Index(ng, name=newgrouper.result_index.name)
 
         elif is_categorical_dtype(self.grouping_vector):
-            # a passed Categorical
-            self._passed_categorical = True
-
-            self.grouping_vector, self._all_grouper = recode_for_groupby(
+            self._cat_grouper = CategoricalGrouper.make(
                 self.grouping_vector, sort, observed
             )
-
+            self.grouping_vector = self._cat_grouper.new_grouping_vector
```

Comment:
why is this the way you would do it? it seems much more natural to have `self.groupbyin_vector` itself returr be a class that can be handled, similar to BaseGrouper (in fact you *should* inherit from BaseGrouper)

Reply:
Thank you for the review @jreback. I'll admit this felt like a kludge. I tried to grok how all the groupby logic interacts, but I mostly failed. There's a BaseGrouper as well as Grouper, but Grouper doesn't inherit from BaseGrouper? What's the difference between a Grouping and a Grouper? A lot of outdated docstrings and lack of typing annotations (and just plain loose typing, eg grouping_vector is sometimes not at all like a vector?) didn't help either.

Even though my choice of name might make it appear otherwise, I wasn't really trying to fit it into any type hierarchy. It is just a container for a few bits of data with some help methods.

I would love to clean this so it makes more sense, but I've been pounding my head on the wall long enough that I think I will need to give up without a bit of guidance. If you're willing to give a summary of how all the classes are intended to interact, then I can try to fix this up. But, I understand you're very busy so if you don't have time to onboard someone new then I get it. Thank you!

Reply:
> There's a BaseGrouper as well as Grouper, but Grouper doesn't inherit from BaseGrouper? What's the difference between a Grouping and a Grouper?

You're right that the hierarchy/naming is confusing.  If you can come up with better names (for the not-public-API items) please go for it.